File: quickSortEasy.pl
#!/usr/bin/perl
#-------------------------------------------------------------------------------
# Quick sort
# Philip R Brenan at gmail dot com, Appa Apps Ltd Inc., 2024
#-------------------------------------------------------------------------------
use v5.38;
use Test::More qw(no_plan);

sub Q{@_<2?@_:(Q(grep$_<$_[@_/2],@_),(grep$_==$_[@_/2],@_),Q(grep$_>$_[@_/2],@_))}  # Quicksort

is_deeply [Q()], [];
is_deeply [Q(2, (1)x4)],        [(1)x4, 2];
is_deeply [Q(qw(6 3 1 4 2 5))], [1..6];
is_deeply [Q(        1..1000)], [1..1000];
is_deeply [Q(reverse 1..1000)], [1..1000];
# Q((1..1000)x10);                                                              # Deep recursion

File: quickSortEasy.py
#-------------------------------------------------------------------------------
# Quick sort in python that crashes on awkward case
# Philip R Brenan at gmail dot com, Appa Apps Ltd Inc., 2025
#-------------------------------------------------------------------------------
def Q(array):
    if len(array) < 2:
        return array
    else:
        p = len(array) // 2
        a = [x for x in array if x <  array[p]]
        P = [x for x in array if x == array[p]]
        b = [x for x in array if x >  array[p]]
        return [*Q(a), *P, *Q(b)]

assert(Q([1,3,2,1,2,3,1]) == [1, 1, 1, 2, 2, 3, 3])

print(Q([j for i in range(10) for j in range(1000)]))                           # Stack overflow
